src/rust/iced-x86/src/decoder/mod.rs

```diff
@@ -853,10 +853,11 @@ impl<'a> Decoder<'a> {
 	#[cfg(has_maybe_uninit)]
 	#[inline]
 	pub fn decode(&mut self) -> Instruction {
-		// Safe, decode_out() initializes the whole thing
-		let mut instruction = unsafe { mem::MaybeUninit::uninit().assume_init() };
-		self.decode_out(&mut instruction);
-		instruction
+		let mut instruction = mem::MaybeUninit::uninit();
+		unsafe {
+			self.decode_out_ptr(instruction.as_mut_ptr());
+			instruction.assume_init()
```

Comment:
@0xd4d I finally got around to checking the assembly, and this change _does_ introduce an unnecessary copy in `Decoder::decode` (when `#[inline]` is removed), but I'm mystified as to why. I can't seem to create a minimal reproduction of the problem, since every time I try, the copy is optimized out as I originally expected.

I created a new crate with a method whose LLVM IR when built without optimizations is exactly the same modulo names of functions and types, but after optimizations, it has no copy. I don't know what's different, but it seems like there must be a problem in the compiler that causes it to miss this optimization in this extremely specific case.

Reply:
I think it's better to just use safe code here (`Instruction::default()`. If someone wants speed, `decode_out()` should be used.

Reply:
Since `decode` is marked inline, I thought it would be a good idea to check the assembly generated for some tests to see how it's optimized after being inlined. I checked (from `decoder::tests::misc_tests`):
* `decode_ip_xxxxxxxxffffffff`
* `decode_multiple_instrs_with_one_instance`
* `decode_with_too_few_bytes_left`
* `position`

The only one of these that copied the `Instruction` after decoding it was `decode_multiple_instrs_with_one_instance`, and I think the copy was unnecessary, but it's hard to say.

Based on this, I think the `MaybeUninit` is worth keeping.

Reply:
Ok, let's keep it.